staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+	queue          k8sworkqueue.RateLimitingInterface
+	workers        int
+
+	// conflicts associates a subnet namespaced name with its conflictsCache.
+	// Always access while holding conflictsMutex.
+	conflicts      map[k8stypes.NamespacedName]*conflictsCache
+	conflictsMutex sync.Mutex
+
+	// staleRVs associates a subnet X's namespaced name for which there was a
+	// successful status update to X's resource version prior to the update.
+	// When a worker begins processing a subnet X, it checks whether X's
+	// resource version matches the resource version in staleRVs[X]. If that's
+	// the case X is stale, i.e. it does not reflect the latest update yet,
+	// hence processing is immediately stopped.
+	// Only access while holding staleRVsMutex.
+	staleRVs      map[k8stypes.NamespacedName]string
+	staleRVsMutex sync.Mutex
+}
+
+func NewValidator(netIfc kosclientv1a1.NetworkV1alpha1Interface,
+	subnetInformer k8scache.SharedInformer,
+	subnetLister netlistv1a1.SubnetLister,
+	queue k8sworkqueue.RateLimitingInterface,
+	workers int) *Validator {
+
+	return &Validator{
+		netIfc:         netIfc,
+		subnetInformer: subnetInformer,
+		subnetLister:   subnetLister,
+		queue:          queue,
+		workers:        workers,
+		conflicts:      make(map[k8stypes.NamespacedName]*conflictsCache),
+		staleRVs:       make(map[k8stypes.NamespacedName]string),
+	}
+}
+
+// Run starts the validator and blocks until stop is closed. This entails
+// starting its Informer and the worker goroutines.
+func (v *Validator) Run(stop <-chan struct{}) error {
+	defer k8sutilruntime.HandleCrash()
+	defer v.queue.ShutDown()
+
+	v.subnetInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    v.OnSubnetCreate,
+		UpdateFunc: v.OnSubnetUpdate,
+		DeleteFunc: v.OnSubnetDelete},
+	)
+
+	go v.subnetInformer.Run(stop)
+	glog.V(2).Infof("Informer run forked.")
+
+	if !k8scache.WaitForCacheSync(stop, v.subnetInformer.HasSynced) {
+		return errors.New("cache failed to sync")
+	}
+	glog.V(2).Infof("Cache synced.")
+
+	// Start workers.
+	for i := 0; i < v.workers; i++ {
+		go k8swait.Until(v.processQueue, time.Second, stop)
+	}
+	glog.V(2).Infof("Launched %d workers.", v.workers)
+
+	<-stop
+	return nil
+}
+
+func (v *Validator) OnSubnetCreate(obj interface{}) {
+	s := obj.(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of creation of %#+v.", s)
+	subnetRef := k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	}
+	v.queue.Add(subnetRef)
+}
+
+func (v *Validator) OnSubnetUpdate(oldObj, newObj interface{}) {
+	oldS, newS := oldObj.(*netv1a1.Subnet), newObj.(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of update from %#+v to %#+v.", oldS, newS)
+	subnetRef := k8stypes.NamespacedName{
+		Namespace: newS.Namespace,
+		Name:      newS.Name,
+	}
+
+	// Process a subnet only if the fields that affect validation have changed.
+	if oldS.Spec.IPv4 != newS.Spec.IPv4 || oldS.Spec.VNI != newS.Spec.VNI {
+		v.queue.Add(subnetRef)
```

Comment:
Move the construction of `subnetRef` into here, it is not needed otherwise.

Reply:
Fixed in 8ac9883.